Refresh Python testing docs (#8102)
microsoft/vscode-docs: docs/python/testing.md

> The [Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) supports testing with Python's built-in [unittest](https://docs.python.org/3/library/unittest.html) framework and [pytest](https://docs.pytest.org/).

Reviewer: Provide a high-level description of what the extension brings: it builds on the built-in testing features in VS Code and provides test discovery, test coverage, and running and debugging tests.

> The [Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) supports testing with Python's built-in [unittest](https://docs.python.org/3/library/unittest.html) framework and [pytest](https://docs.pytest.org/).

Author: ```suggestion
The [Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) builds on the built-in [testing features in VS Code](/docs/debugtest/testing.md) and provides test discovery, test coverage, and running and debugging tests for Python's built-in [unittest](https://docs.python.org/3/library/unittest.html) framework and [pytest](https://docs.pytest.org/).
```

> For a general background on unit testing, read [Unit testing](https://wikipedia.org/wiki/Unit_testing) on Wikipedia. For useful unit test examples, you can review [https://github.com/gwtw/py-sorting](https://github.com/gwtw/py-sorting), a repository with tests for different sorting algorithms.

Reviewer: Can we include a reference to the general testing doc for VS Code (/docs/debugtest/testing.md)?

> For a general background on unit testing, read [Unit testing](https://wikipedia.org/wiki/Unit_testing) on Wikipedia. For useful unit test examples, you can review [https://github.com/gwtw/py-sorting](https://github.com/gwtw/py-sorting), a repository with tests for different sorting algorithms.

Author: We can remove this as well. I don't think it is necessary.
```suggestion
```

## Configure tests

> Once you have the Python extension installed and a Python file open within the editor, a test beaker icon will be displayed on the VS Code Activity bar representing the **Test Explorer** view. When opening the Test Explorer, you will see a **Configure Tests** button if you don't have a test framework enabled. Once you select **Configure Tests**, you will be prompted to select a test framework and a folder containing the tests. If you're using unittest, you will also be asked to select the file glob pattern used to identify your test files.

Reviewer: ```suggestion
Once you have the Python extension installed and a Python file open within the editor, a test beaker icon is displayed on the VS Code Activity Bar representing the **Test Explorer** view. When opening the Test Explorer, you will see a **Configure Tests** button if you don't have a test framework enabled. Once you select **Configure Tests**, you are prompted to select a test framework and a folder containing the tests. If you're using unittest, you are also asked to select the file glob pattern used to identify your test files.
```

Reviewer: Try to avoid future tense unless it's really needed. Makes for a more engaging reading experience.

 > **Note**: A file glob pattern is a defined string pattern that matches file or folder names based on wildcards to then include or not include.

![Configure Python Tests button displayed in the Test Explorer when tests haven't been configured.](images/testing/test-explorer-no-tests.png)

You can configure your tests anytime by using the **Python: Configure Tests** command from the [Command Palette](/docs/getstarted/userinterface.md#command-palette). You can also configure testing manually by setting either `python.testing.unittestEnabled` or `python.testing.pytestEnabled`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/configure/settings.md) documentation. Each framework also has specific configuration settings as described under [Test configuration settings](#test-configuration-settings) for their folders and patterns.

> If you enable pytest, but it is not currently installed in the activated environment, the Python extension will attempt to install it in the background. Furthermore, if both frameworks are enabled, the Python extension will only run `pytest`.

Reviewer: ```suggestion
If you enable pytest, but it is not currently installed in the activated environment, the Python extension attempts to install it in the background. Furthermore, if both frameworks are enabled, the Python extension only runs `pytest`.
```

> **Note**: The minimum supported version of pytest for the python extension is 7.0.0.

## Test discovery

By default, the Python extension attempts to discover tests once you enable a framework. You can also trigger test discovery at any time using the **Test: Refresh Tests** command from the Command Palette.

> > **Tip**: `python.testing.autoTestDiscoverOnSaveEnabled` is set to `true` by default, meaning that test discovery is also performed automatically whenever you add, delete, or update any Python file in the workspace. To disable this feature, set the value to `false`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/editor/settings.md) documentation. You will need to reload the window for this setting to take effect.

Reviewer 2: think we should still include this section: `python.testing.autoTestDiscoverOnSavePattern` setting. Its default value is `**/*.py`.

> > **Tip**: `python.testing.autoTestDiscoverOnSaveEnabled` is set to `true` by default, meaning that test discovery is also performed automatically whenever you add, delete, or update any Python file in the workspace. To disable this feature, set the value to `false`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/editor/settings.md) documentation. You will need to reload the window for this setting to take effect.

Author: ```suggestion
> **Tip**: `python.testing.autoTestDiscoverOnSaveEnabled` is set to `true` by default, meaning that test discovery is also performed automatically whenever you add, delete, or update any Python file in the workspace. To disable this feature, set the value to `false`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/editor/settings.md) documentation. You will need to reload the window for this setting to take effect. For more control over files included in auto test discovery, you can adjust the `python.testing.autoTestDiscoverOnSavePattern` setting which defaults to `**/*.py`.
```

> > **Tip**: `python.testing.autoTestDiscoverOnSaveEnabled` is set to `true` by default, meaning that test discovery is also performed automatically whenever you add, delete, or update any Python file in the workspace. To disable this feature, set the value to `false`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/editor/settings.md) documentation. You will need to reload the window for this setting to take effect. For more control over files included in auto test discovery, you can adjust the `python.testing.autoTestDiscoverOnSavePattern` setting which defaults to `**/*.py`.

Reviewer: ```suggestion
> [!Tip]
> `python.testing.autoTestDiscoverOnSaveEnabled` is set to `true` by default, meaning that test discovery is also performed automatically whenever you add, delete, or update any Python file in the workspace. To disable this feature, set the value to `false`, which can be done either in the Settings editor or in the `settings.json` file as described in the VS Code [Settings](/docs/getstarted/personalize-vscode.md#_configure-settings) documentation. You need to reload the window for this setting to take effect. For more control over files included in auto test discovery, you can adjust the `python.testing.autoTestDiscoverOnSavePattern` setting, which defaults to `**/*.py`.
```

![The VS Code Test Explorer for Python tests](images/testing/test-explorer.png)

> When triggering test discovery directly from the Test Explorer, you can also cancel an ongoing test discovery call. Use the Cancel button, which replaces the Refresh button during discovery.

Reviewer: ```suggestion
When triggering test discovery directly from the Test Explorer, you can also cancel an ongoing test discovery call. Use the **Cancel** button, which replaces the **Refresh** button during discovery.
```

> ![Cancel test discovery button in the Test Explorer.](/images/testing/test-cancel-button.png)

Reviewer: ```suggestion
![Cancel test discovery button in the Test Explorer.](images/testing/test-cancel-button.png)
```

![The VS Code Test Explorer for Python tests](images/testing/test-explorer.png)

When triggering test discovery directly from the Test Explorer, you can also cancel an ongoing test discovery call. Use the Cancel button, which replaces the Refresh button during discovery.

![Cancel test discovery button in the Test Explorer.](/images/testing/test-cancel-button.png)

> If discovery fails (for example, the test framework isn't installed or you have a syntax error in your test file), you'll see an error message displayed in the Test Explorer. The Python extension utilizes error tolerant discovery, meaning if there is an error during discovery that effects some tests but not all, you will see a node of tests and an error node in the Test Explorer. You can check the **Python** output panel to see the entire error message (use the **View** > **Output** menu command to show the **Output** panel, then select **Python** from the dropdown on the right side).

Reviewer 2: there should also be a link to the logs in the error node you see

> If discovery fails (for example, the test framework isn't installed or you have a syntax error in your test file), you'll see an error message displayed in the Test Explorer. The Python extension utilizes error tolerant discovery, meaning if there is an error during discovery that effects some tests but not all, you will see a node of tests and an error node in the Test Explorer. You can check the **Python** output panel to see the entire error message (use the **View** > **Output** menu command to show the **Output** panel, then select **Python** from the dropdown on the right side).

Author: ```suggestion
If discovery fails (for example, the test framework isn't installed or you have a syntax error in your test file), you'll see an error message displayed in the Test Explorer including a link to the logs. The Python extension utilizes error tolerant discovery, meaning if there is an error during discovery that effects some tests but not all, you will see a node of tests and an error node in the Test Explorer. You can check the **Python** output panel to see the entire error message (use the **View** > **Output** menu command to show the **Output** panel, then select **Python** from the dropdown on the right side).
```

> > **Tip**: Unittests can be discovered by pytest, but not vice versa. In the case where testing is configured using unittest but tests are written in pytest, there will be an error message in the Test Explorer.

Reviewer 2: there will not be an error message it will just not find the tests. So if you have pytests but you select unittests then we wont be able to find any tests in the workspace

Author: ![image](https://github.com/user-attachments/assets/13e3c6fd-3332-4416-819b-a698afcd2385)
I see this in the Test Explorer....is this different or unexpected?

> > **Tip**: Unittests can be discovered by pytest, but not vice versa. In the case where testing is configured using unittest but tests are written in pytest, there will be an error message in the Test Explorer.

Author: ```suggestion
```

> [!Note]
> Before running tests with coverage, make sure to install the correct testing coverage package for your project.

Once the coverage run completes, lines are highlighted in the editor for line level coverage. Test coverage results appear as a "Test Coverage" sub-tab in the Test Explorer, which you can also navigate to with **Testing: Focus on Test Coverage View** in Command Palette (`F1)`). On this panel you can view line coverage metrics for each file and folder in your workspace.

![Gif showing running Python tests with coverage.](images/testing/python-coverage.gif)

For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` setting to include your specifications. When the pytest argument `--cov` exists in `python.testing.pytestArgs`, the Python extension will make no additional edits to coverage args, to allow your customizations to take effect. If there is no `--cov` argument found, the extension will add `--cov=.` to the pytest args prior to run to enable coverage at the workspace root.

> > [!Note]
> > Django does not currently support running tests with coverage.

Author: ```suggestion
```

![Test results on a unittest class and in Test Explorer](images/testing/test-results.png)

VS Code also shows test results in the **Test Results** panel.

> ![Test results in the Test Results panel](images/testing/test-result-panel.png)

Reviewer: The image is missing

## Run tests with coverage

> Test coverage is a measure of how much of your code is covered by your tests, which can help you identify areas of your code that are not being fully tested. For more information on test coverage, visit VS Code's [Test Coverage documentation](/docs/editor/testing#_test-coverage.md).

Reviewer: ```suggestion
Test coverage is a measure of how much of your code is covered by your tests, which can help you identify areas of your code that are not being fully tested. For more information on test coverage, visit VS Code's [Test Coverage documentation](/docs/debugtest/testing.md#_test-coverage.md).
```

For finer grain control of your coverage run when using pytest, you can edit the `python.testing.pytestArgs` setting to include your specifications. When the pytest argument `--cov` exists in `python.testing.pytestArgs`, the Python extension will make no additional edits to coverage args, to allow your customizations to take effect. If there is no `--cov` argument found, the extension will add `--cov=.` to the pytest args prior to run to enable coverage at the workspace root.

> > **Note**: Django does not currently support running tests with coverage.

Reviewer: You can use stylized notes as well

```
> [!NOTE]
> put the note here
```

> > **Note**: Django does not currently support running tests with coverage.

Author: ```suggestion
> [!Note]
> Django does not currently support running tests with coverage.
```

If you have more than one configuration entry with `"purpose": ["debug-test"]`, the first definition will be used since we currently don't support multiple definitions for this request type.

## Run tests in parallel

> Support for running tests in parallel with pytest is available through the `pytest-xdist` package. To enable parallel testing:

Author: ```suggestion
Support for running tests in parallel with pytest is available through the `pytest-xdist` package. Visit their [documentation](https://pytest-xdist.readthedocs.io/en/stable/), for more information on how to use `pytest-xdist`.
```

> 1. Open the integrated terminal and install the `pytest-xdist` package. For more details, refer to the [project's documentation page](https://pypi.org/project/pytest-xdist/).
> 
>    **For Windows**
> 
>    ```bat
>    py -3 -m pip install pytest-xdist
>    ```
> 
>    **For macOS/Linux**
> 
>    ```bash
>    python3 -m pip install pytest-xdist
>    ```
> 
> 2. Next, create a file named `pytest.ini` in your project directory and add the content below, specifying the number of CPUs to be used. For example, to set it up for 4 CPUs:
> 
>    ```ini
>     [pytest]
>     addopts=-n4
>    ```
> 
>    Or, if you are using a `pyproject.toml` file
> 
>    ```toml
>     [tool.pytest.ini_options]
>     addopts="-n 4"
>    ```
> 

Author: ```suggestion
```

> If `xdist` is enabled and no worker count is specified in the arguments, the number of workers is automatically optimized by the Python extension based on how many tests were selected in the Test Explorer.

Reviewer 2: I feel like explaining how to install xdist and where to put it in the init file is too verbose / not our job. I would maybe shrink this to just discuss that it can be configured. ALso you can always add -n=4 to the pytest args which is easier than these config files too
